sycl/include/CL/sycl/interop_handle.hpp

```diff
@@ -0,0 +1,117 @@
+//==------------ interop_handle.hpp --- SYCL interop handle ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/access/access.hpp>
+#include <CL/sycl/detail/accessor_impl.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/defines.hpp>
+#include <CL/sycl/detail/pi.hpp>
+
+#include <memory>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+namespace detail {
+class AccessorBaseHost;
+class ExecCGCommand;
+class DispatchHostTask;
+} // namespace detail
+
+template <typename DataT, int Dims, access::mode AccMode,
+          access::target AccTarget, access::placeholder isPlaceholder>
+class accessor;
+
+class interop_handle {
+public:
+  /// Receives a SYCL accessor that has been defined is a requirement for the
+  /// command group, and returns the underlying OpenCL memory object that is
+  /// used by the SYCL runtime. If the accessor passed as parameter is not part
+  /// of the command group requirements (e.g. it is an unregistered placeholder
+  /// accessor), the exception `cl::sycl::invalid_object` is thrown
+  /// asynchronously.
+  template <typename dataT, int dimensions, access::mode accessmode,
+            access::target accessTarget, access::placeholder isPlaceholder>
+  typename std::enable_if<accessTarget != access::target::host_buffer,
+                          cl_mem>::type
+  get_native_mem(const accessor<dataT, dimensions, accessmode, accessTarget,
+                                isPlaceholder> &Acc) const {
+#ifndef __SYCL_DEVICE_ONLY__
+    // employ reinterpret_cast instead of static_cast due to cycle in includes
+    // involving CL/sycl/accessor.hpp
+    auto *AccBase = const_cast<detail::AccessorBaseHost *>(
+        reinterpret_cast<const detail::AccessorBaseHost *>(&Acc));
+    return getMemImpl(detail::getSyclObjImpl(*AccBase).get());
+#else
+    (void)Acc;
+    // we believe this won't be ever called on device side
+    return static_cast<cl_mem>(0x0);
+#endif
+  }
+
+  template <typename dataT, int dimensions, access::mode accessmode,
+            access::target accessTarget, access::placeholder isPlaceholder>
+  typename std::enable_if<accessTarget == access::target::host_buffer,
+                          cl_mem>::type
+  get_native_mem(const accessor<dataT, dimensions, accessmode, accessTarget,
+                                isPlaceholder> &) const {
+    throw invalid_object_error("Getting memory object out of host accessor is "
```

Comment:
You could use a `static_assert` here to prevent users from running code using this.

Reply:
I assume that this is implemented according to [spec](https://github.com/codeplaysoftware/standards-proposals/blob/master/host_task/host_task.md#accessing-low-level-api-functionality-on-sycl-queues):

> ... Trying to obtain the underlying OpenCL object of an accessor with host_buffer target throws an sycl::invalid_object exception.

Reply:
Good point. Thank you for clarifying.